File: Python/Rotting_Oranges.py
"""
https://leetcode.com/problems/rotting-oranges/

Comments included with the code. 

The idea is to go a BFS level order traversal from each rotting orange, then
add 1 to minutes elapsed after each level. Keep a count of how many pure oranges
remained at each point, and when they reach zero, return minutes elapsed.

"""

class Solution(object):
    def orangesRotting(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        dirs = [(0,1),(1,0), (-1,0), (0,-1)]
        q = []
        rows, cols = len(grid), len(grid[0])
        pure_count =0
        # we will do a bfs from a rotten orange

        # first, get the number of pure oranges
        # then add all the rotten oranges in a queue
        for i in range(rows):
            for j in range(cols):
                if grid[i][j]==2:
                    q.append((i,j))
                if grid[i][j]==1:
                    pure_count+=1
        if pure_count==0:
            return 0
        if not q:
            return -1

        # making minutes start from -1, for convenience,
        # as at the next (0th) minute, we will just add
        # the neighbors to the queue, not rot them until the minute after that.
        minutes=-1

        while q:
            next_level = []
            for i, j in q:
                for dx, dy in dirs:
                    x,y = i+dx, j+dy
                    if x>=0 and y>=0 and x<rows and y<cols and grid[x][y]==1:
                        grid[x][y] = 2
                        pure_count-=1
                        next_level.append((x,y))
            q = next_level
            minutes+=1
        return minutes if pure_count==0 else -1

File: C++/RottingOranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {

        bool flag=false;
        int minutes_counter=0;
        if(is_all_rotten(grid)){
            minutes_counter= 0;
            return minutes_counter;
        }
        while(!flag && minutes_counter<=grid.size()*grid[0].size()){
            update(grid,flag);
            minutes_counter++;
        }
        if(flag==false){
            return -1;
        }
        return minutes_counter;


    }
    void update(vector<vector<int>>& grid,bool& flag){
        vector<vector<int>> updated_grid(grid);
        for ( int i=0;i<grid.size();i++){
            for(int j=0;j<grid[i].size();j++){
                if(grid[i][j]==2){//rotten  

                    if(i<grid.size()-1 ){
                        near_rotten_function(updated_grid[i+1][j]);
                    }
                    if(i>0){
                        near_rotten_function(updated_grid[i-1][j]);
                    }
                    if(j<grid[i].size()-1){
                        near_rotten_function(updated_grid[i][j+1]);
                    }
                    if(j>0){
                        near_rotten_function(updated_grid[i][j-1]);
                    }

                }
            }
        }
        grid=updated_grid;
        if(is_all_rotten(grid)){
            flag=true;
        }
    }
    void near_rotten_function(int& cell){
        if (cell==0){
            cell= 0;
            return;
        }
        cell=2;
    }
    bool is_all_rotten(vector<vector<int>>& grid){
        for (int i=0;i<grid.size();i++){
            for(int j=0;j<grid[i].size();j++){
                if (grid[i][j]==1){
                    return false;
                }
            }
        }
        return true;
    }
};
